Bio/SeqIO/__init__.py

```diff
@@ -869,9 +867,15 @@ def index(filename, format, alphabet=None, key_function=None):
         alphabet,
         key_function,
     )
-    return _IndexedSeqFileDict(
-        proxy_class(filename, format), key_function, repr, "SeqRecord"
-    )
+
+    try:
+        random_access_proxy = proxy_class(filename, format)
+    except TypeError:
+        raise TypeError(
+            "Need a string or path-like object for the filename (not a handle)"
+        )
```

Comment:
Does this need ``... from None`` to avoid giving an overly nested exception traceback to the user? Not tested and this one of the changes in modern Python which I've not yet internalised.

Reply:
Yep, I just pushed a commit to fix that. I didn't even know that language feature existed, so I'll definitely be sure to use that in the future. Very useful, thanks for the review!

Reply:
The changes in `SeqIO.index_db` don't `raise` from within `except` blocks, so there shouldn't be a problem there.